### DISCUSSION TITLE
The future of K2forJ4

### DISCUSSION THREAD

**Discussion by the-one-antonio** [eworkers/K2ForJ4#45] [General]
Hi @eworkers 

I want to start by saying that the work you have done here is really amazing.  Kudos, and a sincere thank you.

I wanted to enquire about your intent and ability to continue maintaining K2 into the future.  (We can all likely agree the original developer simply does not have the capacity.)  

I am in a position of trying to decide whether to migrate to J4 with K2forJ4 OR just migrate all content to com_content now.  I suspect there are others in this position.  It is a definite fork in the road because all the current K2->com_content migration tools are designed for J3 only.  For this reason, moving to J4 and K2forJ4 is a risky move.  

I, for one, would love to see this project continue and would be willing to subscribe to it.  But, if you have just released this as band-aid while we wait for Fotis to step up... I think the right decision for me is to take the plunge and move to com_content.  

Perhaps you have not thought this far down the road, but I wanted to open the discussion.  I would love to hear your thoughts if you are willing to share them.

A

**Comment by krmrtrzv** [comment]
@the-one-antonio, Thank you for bringing this dilema that bothers me recently with no resolution on the horizon.
@eworkers, Thousand thanks for your work on K2ForJ4 - huge respect!
Yes, we all K2 users have to make this decision. I would love to be able to stay with K2. It is the most intelligent, effective and smartest extension in Joomla concerning content, but if J5 is EOL and there is no more support for K2ForJ4, then what - a double trouble?

Looking forward the answer on this!

**Comment by AstrocalcVB** [comment]
I'm quite new to Joomla as I previously only worked with WP and before that Drupal 7, but some time ago a friend of my wife asked if I could help with her Joomla site as the guy that used to help her had bailed and didn't want to have anything to do with it anymore.

Anyway, her site was running Joomla 2.something and after some time I was able to bring it up to 3.10.12 but there is was stop due to heavy dependency on K2 (she has 4k+ articles), so this fork really has saved my day as yesterday I was able to upgrade a local copy to J4.4.2 and bow and behold, today I was indeed able to upgrade it further to J5.0.2!

I haven't done any advanced operations but so far everything seems to be functional, but the K2 Quick Icons on the Home Dashboard is distorted, still readable and functional though. I will report back when I have used it a bit more but though to let you know. J5 is not hopeless and thanks for your great work. Keep it up!
V

**Comment by krmrtrzv** [comment]
Hello @eworkers, please give us your vision about the future of K2forJ4. I did the risky move to update to J5 with K2forJ4, and I am happy with it, but now that Jomlaworks announced there is not going to be official K2 development for J>3 anymore, the BIG QUESTION remains - will your fork become officially published, what support do you plan for it, what integration with the old k2 extensions could be maintained for sustainable future.

**Comment by krmrtrzv** [comment]
Hello @eworkers, it's 9 months since this question was published. Please give us your thoughts on your plans for maintaining and developing K2. We need a roadmap, after waiting Fotis for 2 years to hear that K2 is over, it is time for longer perspective plans.

Thank you for saving so many sites from decay with your magnificent intervention!

**Comment by shoulders** [comment]
I would just like clarification on that this project will be continued, I appreciate there will not be constant updates as this requires a lot of manpower 😄 .

I can see there was a commit by @eworkers 2 months ago.

**Comment by the-one-antonio** [comment]
I agree that it is time to address this question.  K2 users were put on ignore by JoomlaWorks for so long.  K2forJ4 is amazing, and the developers have been so kind to provide this - but the long silence is starting to feel like déja vu for users who are already spooked.  

I would like to add that the lack of an "official" K2 means that other developers are reluctant to integrate with K2forJ4.  I've been told by the developers of a couple of extensions that they will not even consider this unless an "official" version is released.  The longer this situation continues, the less relevant K2 becomes... 

Please, please, please... update us, either way... sooner rather than later.  Grazie!

**Comment by the-one-antonio** [comment]
Is anyone there?  This feels an awful lot like déja vu...

**Comment by krmrtrzv** [comment]
Eah, absolutely awful, having in mind, in J3 there was still an option to
use one of the plugins to convert k2 items to core articles. Now w. K2forJ4
and already on J4/5 there are no such options at all. Complete DEADEND

**Comment by shoulders** [comment]
@krmrtrzv maybe the developers of the conversion plugin can add compatibility for K2forJ4 as there should not be a huge difference between them. I am on Joomla3 and need to upgrade it so I am now considering using Joomla articles but my articles have galleries added to them.

**Comment by krmrtrzv** [comment]
developers do not provide support for unofficial extensions, and it is
clear  why should they?
As for the galleries, it is not difficult to change the code of sig pro to
be j4 compatible, but then again it is unofficial version with no future. I
tried the repeatable/subform custom fields and followed the tutorial from
Joomla magazine. It works, you are independent from 3rd party devs, but it
is a lot of manual work and the workflow is very inefficient: uploading
images folder in media manager, then associating the images one by one as
custom fields... and if you got hundred of articles w. galleries...

**Comment by shoulders** [comment]
Some stuff for converting K2 to Joomla articles. These could easily be adapted for K2for J4

**Commercial Plugins**
- https://extensions.joomla.org/extension/migration-a-conversion/joomla-migration/k2-to-content-migration/ 
- https://www.joomlart.com/documentation/joomla-plugins/k2-to-com-content-migration-plugin
- https://extensions.joomla.org/extension/k2-to-content-migration/
- https://joomhelper.com/k2-to-content

**Free Plugins**
- https://github.com/alejoasotelo/migrate-k2-articles-to-joomla

**Tutorials**
- https://medium.com/@thebstring/migrating-from-joomla-3-k2-items-to-joomla-4-articles-42c2a6db890b
- https://www.jelabit.de/migration-k2-en
- https://www.marketingscoop.com/tech/migrating-from-k2-to-joomla-articles-a-step-by-step-guide/

**Comment by krmrtrzv** [comment]
Yes, I know all of them, but they are all valid for J3

**Comment by shoulders** [comment]
> Yes, I know all of them, but they are all valid for J3

but the database should be very similar or might be exactly the same.

**Comment by the-one-antonio** [comment]
***Edit::  Removed this link and comment because we tried another K2 migration tool, spent lots of money on it and wasted so much time trying to get it to work!  I do NOT recommend it!   ****

**Comment by eworkers** [comment]
_First of all, We sincerely apologize for the delayed response in this discussion. Our team has been navigating an extremely challenging situation, living in a brutal war zone for over a year now. Daily life has become a struggle for survival, compounded by the profound loss of many beloved friends, family members, and neighbors. Each day presents new hardships, making it difficult to focus on development while prioritizing our survival and daily life. Despite these challenges, we have managed to allocate time for essential bug fixes whenever possible._

As a reminder, this repository serves as a temporary solution for users who need to utilize K2 with Joomla 4 and 5, especially for those who cannot rely on an official release from the JoomlaWorks team. We have intentionally refrained from adding or removing features and have maintained the original database structure to ensure future compatibility with any official K2 updates from JoomlaWorks. The pace of our work was further influenced by JoomlaWorks' repeated announcements regarding upcoming support updates for Joomla 4 and 5, which initially led us to pause our efforts at multiple occasions.

## What is the future of this repository?
Regarding the future of this repository: JoomlaWorks has indicated that their development efforts are shifting towards Joomla 6. Should they release a version compatible with Joomla 4, 5, and 6, the need for our fork would diminish, as it would have fulfilled its intended purpose. Conversely, if JoomlaWorks continues to show no interest in supporting Joomla 6, we will proceed with enhancing this repository to provide ongoing support for that version (assuming we survived this war).

## Why can’t we publish this fork to the Joomla Extension Directory (JED)?

It is important to reiterate that our fork does not alter K2's core functionality; it solely adds support for Joomla 4 and 5 while addressing existing bugs without breaking the current API or future compatibility with JoomlaWorks releases. The component retains the name com_k2, which is a property of JoomlaWorks on JED. To publish a new version, we would need to rename the component and modify the database tables names, which would lead to breaking compatibility with third-party plugins. Additionally, users transitioning to a differently named fork would lose the ability to upgrade to any future releases from JoomlaWorks.

## Are we planning to release a fork under a different name?
At present, we do not have any projects or clients utilizing the official K2 or this fork. Instead, we are working with an enterprise client who operates a significant news website, which prompted us to create a custom fork of K2ForJ4. This project allows us the flexibility to add and remove features as needed. We have implemented substantial refactoring, performance improvements, and adaptations to the MVC structure, aligning it more closely with the newer MVC standards introduced in Joomla 3.10 and 4.0.

This custom version features a different API and will not be compatible with any existing third-party K2 addons. However, it should not require excessive effort on the part of third-party developers to adapt their addons for compatibility with our new fork. We plan to share our developments on GitHub, particularly when we add a migration tool from K2 to the new component (currently requires that users renames/copy DB tables and copy/rename media folders to a new directory), though a decision on submitting this forked component to JED has yet to be made. though a decision on submitting this forked component to JED has yet to be made.

**Comment by shoulders** [comment]
@eworkers thank you very much for this update and it is good to know the project is still live and when needed, patches and updates will be made, if time allows for it. It is also good to know that the database is still compatible with the "Official" K2 and that users of the **K2forJ4** version will not be "stuck" without an upgrade path for future versions of Joomla.

**Comment by krmrtrzv** [comment]
Dear @eworkers <https://github.com/eworkers>,
Thank you so much for your clear and highly needed answer to the question
of the future prospects of K2ForJ4!
I am sorry to hear you pass through the nightmares of war, breaking apart
the world that has already been on the brink of climate catastrophe. I hope
the war is over soon and in a just manner for all the suffering from its
atrocities!
As for the complicated relations between Joomlaworks official K2 and your
fork - that is completely comprehensible and I admire your clear approach
to that.
I just wanted to point on what you are saying concerning Joomlaworks plans
to make K2 compatible to Joomla 6, which I have not encountered anywhere in
public.
On May 30, 2024 Joomlaworks announced in bold manner that "K2 will not be
made available for Joomla 4/5 - change of course" on their blog <
https://www.joomlaworks.net/blog/item/314-k2-will-not-be-made-available-for-joomla-4-5-change-of-course>.
They announced that the plan is "to go from K2 (a "half" CMS so to speak)
to a full CMS" - a fork to Joomla!!! that is built with K2 as a core
content component (Joomla 6 is not mentioned in that announcement) - a
giant effort and big scale project, as far as my modest knowledge could
tell, requiring a large team and investment, but since that single
announcement was made 6 months ago was followed by complete sillence and
considering the last 4 years of promises not fulfilled by Joomlaworks, I
have my doubts the plan is feasible and would not wait another four (or
more) years facing JW lack of transparency.

I wonder if that would change your course and tactics?

Take care!

**Comment by the-one-antonio** [comment]
Dear @eworkers 

I believe I can speak for many here when I say that I am so sorry about what you are going through.  It is utterly heartbreaking to hear.  It also makes your work here even more remarkable - I apologize most sincerely for requesting updates, in my total ignorance of what you are experiencing.  

There are many people who have benefited from your efforts and I think you will find financial support here if you choose to continue and make your work into a paid extension.  I, for one, would like to support you if there is a way to do so.  

Please accept my sincere thanks - and please know we will be thinking about you, very much hoping for peace to return to your land and your lives.

**Comment by eworkers** [comment]
Thank you all for checking in and for your patience — we truly appreciate it. You can read our full reply (including roadmap and release timing) [here](https://github.com/eworkers/K2ForJ4/discussions/94#discussioncomment-15778760)
